airbyte-persistent-queue/src/main/java/io/airbyte/persistentqueue/CloseableInputQueue.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Airbyte
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.persistentqueue;
+
+import java.util.Queue;
+
+/**
+ * An autocloseable queue with a hook for signal when no more records will written to the queue.
+ * Intended for use with queues that are accessed across multiple threads and that contain
+ * connections or resources that should be cleaned up after use. See
+ * {@link AbstractCloseableInputQueue} for an example of the state machine that this interface
+ * provides.
+ *
+ * @param <E>
+ */
+public interface CloseableInputQueue<E> extends Queue<E>, AutoCloseable {
```

Comment:
I'm not sure how valuable it is to extent the `Queue` interface. The main reason that I'm questioning it, is that it would be nice to be able to have some of these messages throw IOException. Inclined to keep it as is, but curious if anyone feels differently.